contracts/utils/cryptography/MultiSignerERC7913Weighted.sol

```diff
@@ -0,0 +1,171 @@
+
+/**
+ * @dev Extension of {MultiSignerERC7913} that supports weighted signatures.
+ *
+ * This contract allows assigning different weights to each signer, enabling more
+ * flexible governance schemes. For example, some signers could have higher weight
+ * than others, allowing for weighted voting or prioritized authorization.
+ *
+ * Example of usage:
+ *
+ * ```solidity
+ * contract MyWeightedMultiSignerAccount is Account, MultiSignerERC7913Weighted, Initializable {
+ *     constructor() EIP712("MyWeightedMultiSignerAccount", "1") {}
+ *
+ *     function initialize(bytes[] memory signers, uint256[] memory weights, uint256 threshold) public initializer {
+ *         _addSigners(signers);
+ *         _setSignerWeights(signers, weights);
+ *         _setThreshold(threshold);
+ *     }
+ *
+ *     function addSigners(bytes[] memory signers) public onlyEntryPointOrSelf {
+ *         _addSigners(signers);
+ *     }
+ *
+ *     function removeSigners(bytes[] memory signers) public onlyEntryPointOrSelf {
+ *         _removeSigners(signers);
+ *     }
+ *
+ *     function setThreshold(uint256 threshold) public onlyEntryPointOrSelf {
+ *         _setThreshold(threshold);
+ *     }
+ *
+ *     function setSignerWeights(bytes[] memory signers, uint256[] memory weights) public onlyEntryPointOrSelf {
+ *         _setSignerWeights(signers, weights);
+ *     }
+ * }
+ * ```
+ *
+ * IMPORTANT: When setting a threshold value, ensure it matches the scale used for signer weights.
+ * For example, if signers have weights like 1, 2, or 3, then a threshold of 4 would require at
+ * least two signers (e.g., one with weight 1 and one with weight 3). See {signerWeight}.
+ */
+abstract contract MultiSignerERC7913Weighted is MultiSignerERC7913 {
+    using EnumerableSetExtended for EnumerableSetExtended.BytesSet;
+    using SafeCast for uint256;
+
+    // Invariant: sum(weights) >= threshold
+    uint128 private _totalWeight;
+
+    // Mapping from signer ID to weight
+    mapping(bytes32 signedId => uint256) private _weights;
+
+    /// @dev Emitted when a signer's weight is changed.
+    event ERC7913SignerWeightChanged(bytes indexed signer, uint256 weight);
+
+    /// @dev Thrown when a signer's weight is invalid.
+    error MultiSignerERC7913WeightedInvalidWeight(bytes signer, uint256 weight);
+
+    /// @dev Thrown when the threshold is unreachable.
+    error MultiSignerERC7913WeightedMismatchedLength();
+
+    /// @dev Gets the weight of a signer. Returns 0 if the signer is not authorized.
+    function signerWeight(bytes memory signer) public view virtual returns (uint256) {
+        return Math.ternary(isSigner(signer), _signerWeight(signer), 0);
+    }
+
+    /// @dev Gets the total weight of all signers.
+    function totalWeight() public view virtual returns (uint256) {
+        return _totalWeight; // Doesn't need Mat.max because it's incremented by the default 1 in `_addSigners`
```

Comment:
Math

Reply:
Addressed in https://github.com/OpenZeppelin/openzeppelin-community-contracts/pull/125/commits/e5bfa3e9905bf0c68b2d51b165a66654b36c28c5